In [5]:
#load data
from sklearn.datasets import fetch_california_housing
import pandas as pd

#load housing dataset
housing = fetch_california_housing()

#create dataframe
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

#initial exploration
df = X.copy()
df["med_house_value"] = y

display(df.head())

#print names
print("Feature Names:")
print(X.columns)

#check for missing values
print("\nMissing Values per Column:")
print(df.isnull().sum())

#summary statistics
display(df.describe())

from sklearn.model_selection import train_test_split

#80% training, 20% testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("training set shape:", X_train.shape)
print("test set shape:", X_test.shape)

from sklearn.linear_model import LinearRegression

#initialize and train model
model = LinearRegression()
model.fit(X_train, y_train)

#predict on test set
y_pred = model.predict(X_test)

#view first 5 predictions
print("first 5 predictions:", y_pred[:5])
print("first 5 actual values:", y_test.values[:5])

#MSE, RMSE, R^2
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

mse = mean_squared_error(y_test, y_pred)
print("MSE:", mse)
rmse = np.sqrt(mse)
print("RMSE:", rmse)
r2 = r2_score(y_test, y_pred)
print("R² Score:", r2)

#select features
#I am choosing "MedInc," "AveRooms," and "Latitude" because median income is the single strongest predictor of house value (wealthy people can buy more expensive homes)
#I chose AveRooms because the more number of rooms in a home, the more people can live there --> the higher the home can be listed at a certain price
#Finally, I chose latitude because I know southern California can get pretty expenseive in/near the big cities, whereas the North is more ambiguous

#train and test model
selected_features = ['MedInc', 'AveRooms', 'Latitude']
X_simple = X[selected_features]
X_train_s, X_test_s, y_train_s, y_test_s = train_test_split(
    X_simple, y, test_size=0.2, random_state=42
)
simple_model = LinearRegression()
simple_model.fit(X_train_s, y_train_s)

#compare
y_pred_s = simple_model.predict(X_test_s)
mse_s  = mean_squared_error(y_test_s, y_pred_s)
rmse_s = np.sqrt(mse_s)
r2_s   = r2_score(y_test_s, y_pred_s)

print("MSE (simplified):", mse_s)
print("RMSE (simplified):", rmse_s)
print("R² Score (simplified):", r2_s)

print("\nFull vs. Simplified Comparison")
print(f"Full Model    → MSE: {mse:.4f}  RMSE: {rmse:.4f}  R²: {r2:.4f}")
print(f"Simple Model  → MSE: {mse_s:.4f}  RMSE: {rmse_s:.4f}  R²: {r2_s:.4f}")

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,med_house_value
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


Feature Names:
Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude'],
      dtype='object')

Missing Values per Column:
MedInc             0
HouseAge           0
AveRooms           0
AveBedrms          0
Population         0
AveOccup           0
Latitude           0
Longitude          0
med_house_value    0
dtype: int64


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,med_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


training set shape: (16512, 8)
test set shape: (4128, 8)
first 5 predictions: [0.71912284 1.76401657 2.70965883 2.83892593 2.60465725]
first 5 actual values: [0.477   0.458   5.00001 2.186   2.78   ]
MSE: 0.5558915986952436
RMSE: 0.7455813830127759
R² Score: 0.5757877060324513
MSE (simplified): 0.6941883210287301
RMSE (simplified): 0.833179645111863
R² Score (simplified): 0.47025063735398676

Full vs. Simplified Comparison
Full Model    → MSE: 0.5559  RMSE: 0.7456  R²: 0.5758
Simple Model  → MSE: 0.6942  RMSE: 0.8332  R²: 0.4703


Interpretation Questions:

Part 2:
- The R^2 value tells us the proportion of variance in median house values explained by the model. So, an R² of .58 means roughly 60% of the variability is captured by the linear model, which is a moderate linear fit.

- The features that seem to have the strongest feedback based on the on prediction's based on the model's coefficients are MedInc mostly, as well as latitiude and longitude as well.

- The RMSE is roughly $0.72, in units of $100k, meaning predictions are off by about $72,000 on average. The scatter between predicted and actual values is noticeable, so the model captures the general trend but struggles with the higher outliers, which checks out though for a simple linear model.

Part 3:
- The simplified model achieves an R² of about .47 vs .58 for the full model  a kind of big drop. This tells us the full model explains noticeably more variance in house prices. The 5 removed features (together) contribute significant predictive power that the simplified model loses.

- For me personally, I do not think that I would use this model. If I am trying to make a maningful prediction, but I know that the simplified model is "less correlated" by almost eleven percent, I would stay awyay. This being said, if I needed a quick prediction for a price, and this is the model that I had access to at the time, I would not have a problem using it for a simple value that I want to know (although I would probably check it again later).